server/src/main/java/org/elasticsearch/search/internal/InternalTimeoutException.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.search.internal;
+
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.query.QuerySearchResult;
+import org.elasticsearch.search.query.SearchTimeoutException;
+
+/**
+ * Base exception thrown whenever a search timeout occurs. It can only be subclassed and created effectively via
+ * {@link ContextIndexSearcher#throwTimeExceededException()}. This more generic base exception can be used to catch timeout errors
+ * in the different search phases and handle timeouts appropriately
+ */
+public abstract sealed class InternalTimeoutException extends RuntimeException permits ContextIndexSearcher.TimeExceededException {
```

Comment:
This layer of indirection is a little convoluted, I wonder if it's overkill. Its purpose is to not have search components depend on `ContextIndexSearcher.TimeExceededException` directly as that seems like an anti-pattern. But consumers like RescorePhase and FetchPhase (and soon DFSPhase?) will need to catch such timeout exception to handle it correctly.

Reply:
Not sure I fully understand the motivation here. 

But aren't we essentially still leaking the `ContextIndexSearcher`? We want this to only be thrown by the `ContextIndexSearcher` and it will never be thrown by any other searcher.
We also expose the `ContextIndexSearcher` via `org.elasticsearch.search.internal.SearchContext#searcher` explicitly and use that in search phases.
It seems we need to have the knowledge about relation between the timeout exception and that it's only ever thrown by the `ContextIndexSearcher` in caller code in some form, what are we trying to hide? :)

Reply:
I simplified this and removed the additional exception that was just swiping things under the rug :)